libethash-cuda/CUDAMiner.cpp

```diff
@@ -509,17 +492,33 @@ void CUDAMiner::search(
 				buffer[0] = 0;
 				if (found_count > (SEARCH_RESULT_BUFFER_SIZE - 1))
 					found_count = SEARCH_RESULT_BUFFER_SIZE - 1;
-				for (unsigned int j = 0; j < found_count; j++)
-					nonces[j] = nonce_base + buffer[j + 1];
+				for (unsigned int j = 1; j <= found_count; j++)
+					nonces[j] = nonce_base + buffer[j];
 			}
 		}
 		run_ethash_search(s_gridSize, s_blockSize, m_sharedBytes, stream, buffer, m_current_nonce, m_parallelHash);
 		if (m_current_index >= s_numStreams)
 		{
 			if (found_count)
-				found(nonces, found_count, w);
-			searched(batch_size);
-			exit = cuda_shouldStop();
+			{
+				for (uint32_t i = 1; i <= found_count; i++)
+				{
+					Result r = EthashAux::eval(w.seed, w.header, nonces[i]);
+					if (r.value < w.boundary)
+						farm.submitProof(Solution{nonces[i], r.mixHash, w.header, w.seed, w.boundary, w.job, w.job_len, m_abort});
+					else
+					{
+						farm.failedSolution();
+						cwarn << "GPU gave incorrect result!";
+					}
+				}
+			}
+			addHashCount(batch_size);
+			if (m_abort || shouldStop())
```

Comment:
The `shouldStop()` also has some kind of "abort" flag. I'm not sure we need both.

Reply:
Yeah, that's a little confusing. That should stop is the Farm's shouldStop, not the miner's. It's only used for terminating operations.